### ISSUE TITLE
(My crazy?) Plan to re-unite 0.23 and 1.0

### ISSUE THREAD

**Issue by armanbilge** [http4s/http4s#5255]
We've already made big progress towards this via #5252 and #5241 and will be helped along by #5226 (see also #5044).

My assumptions:
1. Once upon a time, 0.23 and 1.0 were the same, until I completely destroyed 1.0 with Scala.js.
2. Besides JS, the diff between 0.23 and 1.0 at the moment is _very_ small (does anyone have a list?).
3. 0.23 will be merging into 1.0 for years to come.
4. "Platformed" files (split across JVM/JS) are the root of the worst merge conflicts.
5. Most of the file-splitting I did to cross-compile to JS was due to lack of cross-compiling crypto and fs2-io. Both have since been addressed, and all these files can be glued back together.
6. Off the top of my head, the biggest breaking changes actually needed to cleanly cross to JS are:
  * Some `Async` constraints instead of `Sync`, but only for crypto (and only needed for browsers).
  * Using ip4s for the `Server` address, rather than a JVMism.

What this means (bear with me): we can probably cross-compile 0.23 core, circe, and client for JS (minus crypto support in browsers) with minimal effort i.e. just by fiddling with the build. I'm not saying we _should_ do this, although it would:
* tighten the diff in the build configuration between 0.23 and 1.0;
* force 0.23 feature development to target JS as well, to avoid painful merges into 1.0;
* certainly be nice to have a stable release with JS support.

On the other hand, this would push us even further away from 0.22.

In any case: what I'd like to do (if it's not too insane) is "re-cut" 1.0 from the updated 0.23. The thinking is, instead of glueing files back together in 1.0, we have a much better chance to quash future merge conflicts by re-copying the affected files directly out of 0.23 (of course, being very careful not to overwrite any of the nice non-JS things added to 1.0 since). The more we can do to make 0.23 like 1.0 the nicer the outcome will be. 

Anyway, just a (crazy) idea and understandably the ship may have sailed on some of this. But if it's still an option there's some exciting things looming for 1.0-only that once landed will make this more difficult.

Finally, apologies for how time/energy-consuming this JS stuff has been, I _really_ hope things will be smoother from here.

**Comment by rossabaker** [comment]
> Once upon a time, 0.23 and 1.0 were the same, until I completely destroyed 1.0 with Scala.js.

True, but someone had to be first!

> Besides JS, the diff between 0.23 and 1.0 at the moment is very small (does anyone have a list?).

It's not perfect, but this is a reasonable overview:

```sh
git log ^series/0.23 main --grep='#' --format='oneline'
```

Temporarily turning on MiMa against the latest 0.23 is another tool.  Only a rigorous diff is certain, though.

> 0.23 will be merging into 1.0 for years to come.

I hope that's not true, but historically about a year after 1.0 is released, and we don't have a concrete timeline for 1.0.  A Great Schism between 0.23 and 1.0 is going to make that process harder than anything we've done before, though, and we're going to have to think about what that means.

> "Platformed" files (split across JVM/JS) are the root of the worst merge conflicts.

Git is doing a really good job merging into the renamed locations, except when it's not.  And it's easy to get duplicated files and tests that don't run.  The fear has been worse than the labor.

> Most of the file-splitting I did to cross-compile to JS was due to lack of cross-compiling crypto and fs2-io. Both have since been addressed, and all these files can be glued back together.

Yep, and this is tremendous.

> What this means (bear with me): we can probably cross-compile 0.23 core, circe, and client for JS (minus crypto support in browsers) with minimal effort i.e. just by fiddling with the build.

We are in a much better place than we were last week, but we would need be sure it's clear how to clone and develop the project, and we need to be sure the bus factor is sufficient for the production line.  @ChristopherDavenport and @cquiroz are the only other two committers who have shown much aptitude for Scala.js.  Can they continue to step up?  We can also buy more time to learn and attract new contributors if we issue no binary compatibility guarantees for the JavaScript.

> Some Async constraints instead of Sync, but only for crypto (and only needed for browsers).
> Using ip4s for the Server address, rather than a JVMism.

These are going to be tricky on 0.23, or is that why you're being modest in the modules?

> In any case: what I'd like to do (if it's not too insane) is "re-cut" 1.0 from the updated 0.23. The thinking is, instead of glueing files back together in 1.0, we have a much better chance to quash future merge conflicts by re-copying the affected files directly out of 0.23

I'd need to understand better what "re-cut" means in terms of git.

**Comment by armanbilge** [comment]
Thanks for your thoughts!

> I hope that's not true, but historically about a year after 1.0 is released, and we don't have a concrete timeline for 1.0.

I hope so too, but the discussion in #5245, paired with the lack of concrete timeline for the complex changes going into 1.0, makes me worry we are in for a long haul.

> A Great Schism between 0.23 and 1.0 is going to make that process harder than anything we've done before, though, and we're going to have to think about what that means.

I'd like to remove JS-related headaches from this process as much as possible, whatever ends up being the best strategy to do so. That's what this issue is really about, I guess.

> These are going to be tricky on 0.23, or is that why you're being modest in the modules?

Exactly. I think core, (ember-)client, and circe are the most important modules for JS (it gets us everything we need for http4s-dom and typelevel/feral), so we can live without server, and browsers will just have to get along without crypto or use a 1.0 milestone instead.

> We are in a much better place than we were last week, but we would need be sure it's clear how to clone and develop the project, and we need to be sure the bus factor is sufficient for the production line.

It was the brief affair with browsers and the external servers that made the dev experience terrible. Now we are back to a place where all you need is Node 16 and yarn installed and everything should work out-of-the-box exactly like it does for CE, fs2, vault, keypool, and many Davenverse projects. I'll make a long-overdue PR for the contributing docs with this information.

> if we issue no binary compatibility guarantees for the JavaScript.

Now that the DOM modules moved out, effectively all the JS APIs in http4s are shared with JVM and thus already covered by the usual binary compatibility guarantees. The only interface that is unique to JS is the node-serverless module and I'm not in any rush to make any binary compatibility guarantees about it yet :)

> @ChristopherDavenport and @cquiroz are the only other two committers who have shown much aptitude for Scala.js

Both @cquiroz and @rpiaggio have stepped up as maintainers for http4s-dom. And also @isomarcte did some really thorough reviews. Hopefully we'll attract additional maintainers as serverless projects like typelevel/feral ramp up.

> I'd need to understand better what "re-cut" means in terms of git.

Well, the most aggressive execution would be to declare the current main branch "cursed", branch a new main off of 0.23 and re-do everything "right". But I assume that's way too crazy 😆 

A more reasonable approach would be to emulate this by manually working through the diff between 0.23 and main and making changes specifically aiming to minimize the diff.

> Git is doing a really good job merging into the renamed locations, except when it's not. And it's easy to get duplicated files and tests that don't run. The fear has been worse than the labor.

This is true for many files, but I've also had some really challenging merges with certain files. And it's very easy for casual JVMisms to sneak into 0.23 sources, which we don't realize until the merge hits 1.0, where we can probably fix, but then the diff has unnecessarily increased between 0.23 and 1.0, making any future merges just a little trickier.

Actually, this happened in my merge today because someone had the audacity to use JDK's native `URLEncoder`. The ideal way to fix issues like this is to go back and fix it in 0.23 and then merge it back up, to keep the diff small. Or even better to anticipate it in the first place.

**Comment by bplommer** [comment]
Just a random thought, but is there any reason why splits into platform-specific source files can't be backported to 0.22/0.23 for subsequent ease of forward merging, even when there's only a jvm file in those versions?

**Comment by armanbilge** [comment]
> This is probably a terrible idea but what about cross-building 0.23 for scala.js, with the js version not being published and all the js implementations being stubbed out with `???`?
> 
> EDIT: that second suggestion isn't not going to work, as it still relies on dependencies being available for js.

Actually, that's what part of what I'm suggesting. The 0.23 dependencies (which are shared with 1.0) _are_ now available for JS. And the amount of JS specific (i.e. non-shared) code in 1.0 is very very small (it's been moved upstream into fs2.io and http4s-crypto). If you have a look at the JS sources for core you can see it's mostly just empty files: https://github.com/http4s/http4s/tree/main/core/js/src/main/scala/org/http4s

That's why I'm pretty sure we've arrived to a place where we can "flip the switch" on several 0.23 modules and they should already be publishable for JS (or, very nearly so).

> Just a random thought, but is there any reason why splits into platform-specific source files can't be backported to 0.22/0.23 for subsequent ease of forward merging, even when there's only a jvm file in those versions?

It could work, the hard part is policing these files in 0.22/0.23 since there'll be no tests/compiler guidance about what should or shouldn't be platformed. The better news is that particularly after #5266 merges I think we can get rid of nearly all these split files actually.

**Comment by cquiroz** [comment]
Is the idea no to have js specific code? I'm thinking of a minimized MimeDb, is that out of the question

**Comment by armanbilge** [comment]
> Is the idea no to have js specific code?

The main idea has been to minimize the diff between 0.22/0.23/1.0. Hiding JVM/JS-specific code behind shared APIs such as http4s-crypto has so far been the best strategy towards this.

> I'm thinking of a minimized MimeDb, is that out of the question

👍 I'd really like to do this as well, but I'm still not sure what the strategy would be so that we could still keep the entire MimeDB for Node.js. See my comment in https://github.com/http4s/http4s/issues/5113#issuecomment-906834998.

**Comment by rossabaker** [comment]
I think we can revisit the notion of the MimeDB, given that we're creating constants for proprietary formats for business that went under a decade ago. The bar to get into that registry wasn't high.  It's neat that we're adhering to standards, but rearranging how we adhere to it is definitely on the table.

**Comment by armanbilge** [comment]
Executed in #5298 and #5303.
